lib/content-patcher/patchers/class-patcher-abstract.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace NewspackContentConverter\ContentPatcher\Patchers;
+
+use NewspackContentConverter\ContentPatcher\PatcherInterface;
+
+/**
+ * Class PatcherAbstract, containing common functionality.
+ *
+ * @package NewspackContentConverter\ContentPatcher\Patchers
+ */
+abstract class PatcherAbstract implements PatcherInterface {
+
+	/**
+	 * @var array Array of regex expressions -- keys are self explanatory.
+	 *      'match_html_element'
+	 *      'match_blocks_element'
+	 *      'match_attribute_value' - has an extra group, and matches the targeted HTML element attribute's value
+	 *      'replace_blocks_element'
+	 */
+	protected $patterns = [];
+
+	/**
+	 * Runs a preg_match_all with PREG_OFFSET_CAPTURE.
+	 *
+	 * @param string $pattern Pattern.
+	 * @param string $subject Subject.
+	 *
+	 * @return array|null preg_match_all with PREG_OFFSET_CAPTURE matches, or null if no matches.
+	 */
+	protected function match_all_elements( $pattern, $subject ) {
+	    $preg_match_all_result = preg_match_all( $pattern, $subject, $matches, PREG_OFFSET_CAPTURE );
+
+	    return ( false === $preg_match_all_result || 0 === $preg_match_all_result ) ? null : $matches;
+	}
+
+	/**
+	 * Validates the results of match searches.
+	 *
+	 * @param array $matches_html   Regex matches in HTML source.
+	 * @param array $matches_blocks Regex matches in blocks content.
+	 *
+	 * @return bool Do the matches correspond.
+	 */
+	protected function validate_html_and_block_matches( $matches_html, $matches_blocks ) {
```

Comment:
Several cases come to my mind. First is that a different number is found. Second is that an element could be converted differently under different contexts (eg. when standalone HTML element, or nested within a certain structure then converted differently because of that structure, or simply because of a bug in conversion). These are wide shots, but these kind of checks make room for custom logic now. We might remove them, but I'd like to see a few patchers written before doing more refactoring.